### PR TITLE
Added Web3 Staking Dapp

### LINKED ISSUE #550: [Add a Play]: Web3 Staking Dap
Thank You! 

## What's the Play about?
It is a Staking Dapp where people will get interest based on the amount they stake after a specific period of time.

## What ReactJS concept will be used to create this play?
React hooks like useState, useEffect

## Are you willing to take it up for implementation?
yes

### PR DESCRIPTION
**Pull request by nirban256**
# Description

It is a Staking Dapp where people will get interest based on the amount they stake after a specific period of time. I have added some dependencies namely hardhat and ethers in order to build the Dapp (Decentralized Application).

Fixes #550 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

# Checklist:

- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by nirban256** [comment]
@atapas pls review the pr

File: src/plays/web3-staking-dapp/artifacts/build-info/5faf9c6c2f8753632f3c7c13937cd196.json [hunk 1]
```diff
@@ -0,0 +1 @@
...9 JUMPI PUSH2 0x1358 PUSH2 0x147E JUMP JUMPDEST JUMPDEST DUP3 DUP3 DIV SWAP1 POP SWAP3 SWAP2 POP POP JUMP JUMPDEST PUSH1 0x0 PUSH2 0x136F DUP3 PUSH2 0x13FC JUMP JUMPDEST SWAP2 POP PUSH2 0x137A DUP4 PUSH2 0x13FC JUMP JUMPDEST SWAP3 POP DUP2 PUSH32 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF DIV DUP4 GT DUP3 ISZERO ISZERO AND ISZERO PUSH2 0x13B3 JUMPI PUSH2 0x13B2 PUSH2 0x144F JUMP JUMPDEST JUMPDEST DUP3 DUP3 MUL SWAP1 POP SWAP3 SWAP2 POP POP JUMP JUMPDEST PUSH1 0x0 PUSH2 0x13C9 DUP3 PUSH2 0x13DC JUMP JUMPDEST SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x0 DUP2 ISZERO ISZERO SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x0 PUSH20 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF DUP3 AND SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x0 DUP2 SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH1 0x0 PUSH2 0x1411 DUP3 PUSH2 0x13FC JUMP JUMPDEST SWAP2 POP PUSH32 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF DUP3 EQ ISZERO PUSH2 0x1444 JUMPI PUSH2 0x1443 PUSH2 0x144F JUMP JUMPDEST JUMPDEST PUSH1 0x1 DUP3 ADD SWAP1 POP SWAP2 SWAP1 POP JUMP JUMPDEST PUSH32 0x4E487B7100000000000000000000000000000000000000000000000000000000 PUSH1 0x0 MSTORE PUSH1 0x11 PUSH1 0
```
Comment (atapas) [inline src/plays/web3-staking-dapp/artifacts/build-info/5faf9c6c2f8753632f3c7c13937cd196.json:1]: Is this file intended?

File: src/plays/web3-staking-dapp/artifacts/contracts/Staking.sol/Staking.dbg.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (atapas) [inline src/plays/web3-staking-dapp/artifacts/contracts/Staking.sol/Staking.dbg.json:1]: Is this file needed?

File: src/plays/web3-staking-dapp/styles.css [hunk 3]
```diff
@@ -0,0 +1,214 @@
+/* Navbar */
```
Comment (atapas) [inline src/plays/web3-staking-dapp/styles.css:1]: Remove commented code

**Review by atapas — CHANGES_REQUESTED** [review]
I have given some initial comments.

@Sachin-chaurasiya will do deep review.

File: src/plays/web3-staking-dapp/styles.css (hunk 3 above)
Comment (nirban256) [reply src/plays/web3-staking-dapp/styles.css:1]: ok

File: src/plays/web3-staking-dapp/artifacts/build-info/5faf9c6c2f8753632f3c7c13937cd196.json (hunk 1 above)
Comment (nirban256) [reply src/plays/web3-staking-dapp/artifacts/build-info/5faf9c6c2f8753632f3c7c13937cd196.json:1]: I think it will be better if we keep this file

File: src/plays/web3-staking-dapp/artifacts/contracts/Staking.sol/Staking.dbg.json (hunk 2 above)
Comment (nirban256) [reply src/plays/web3-staking-dapp/artifacts/contracts/Staking.sol/Staking.dbg.json:1]: I think for a web3 dapp we should keep all the files related to it so I have kept the files

File: src/plays/web3-staking-dapp/Readme.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+Update your implementation idea and details here
+
+## Consideration
+
+Update all considerations(if any)
+
+## Resources
+
+Update external resources(if any)
```
Comment (Sachin-chaurasiya) [inline src/plays/web3-staking-dapp/Readme.md:27]: Can we update these details as well?

File: src/plays/web3-staking-dapp/Web3StakingDapp.js [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+import maticLogo from "./images/polygon-matic-logo.svg";
+import { FaRupeeSign } from 'react-icons/fa';
+
+// WARNING: Do not change the entry componenet name
+function Web3StakingDapp(props) {
+
+  // Your Code Start below.
+
+  const ContractAddress = '0x12163B070B97f06F5061D93164D960bbFCfdf965';
+
```
Comment (Sachin-chaurasiya) [inline src/plays/web3-staking-dapp/Web3StakingDapp.js:16]: Move this to constant.

File: src/plays/web3-staking-dapp/Web3StakingDapp.js [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+  // frontend states
+  const [provider, setProvider] = useState(undefined);
+  const [account, setAccount] = useState(undefined);
+  const [contract, setContract] = useState(undefined);
+  const [accountAddress, setAccountAddress] = useState(undefined);
```
Comment (Sachin-chaurasiya) [inline src/plays/web3-staking-dapp/Web3StakingDapp.js:21]: We don't need to provide the default value as `undefined` as it will be undefined by default.
```suggestion
  const [provider, setProvider] = useState();
  const [account, setAccount] = useState();
  const [contract, setContract] = useState();
  const [accountAddress, setAccountAddress] = useState();
```

File: src/plays/web3-staking-dapp/Web3StakingDapp.js [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+  // assets
+  const [assetId, setAssetId] = useState([]);
+  const [assets, setAssets] = useState([]);
+
+  // staking
+  const [stakeModal, setStakeModal] = useState(false);
+  const [stakingLength, setStakingLength] = useState(undefined);
+  const [stakingPercent, setStakingPercent] = useState(undefined);
```
Comment (Sachin-chaurasiya) [inline src/plays/web3-staking-dapp/Web3StakingDapp.js:30]:
```suggestion
  const [stakingLength, setStakingLength] = useState();
  const [stakingPercent, setStakingPercent] = useState();
```

File: src/plays/web3-staking-dapp/Web3StakingDapp.js [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+    const onload = async () => {
+      const provider = new ethers.providers.Web3Provider(window.ethereum);
+      setProvider(provider);
+
+      const contract = new ethers.Contract(
+        ContractAddress,
+        artifact.abi
+      )
+      setContract(contract);
+    }
```
Comment (Sachin-chaurasiya) [inline src/plays/web3-staking-dapp/Web3StakingDapp.js:47]: Can we Move this function out of the useEffect?

File: src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json [hunk 9]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (Sachin-chaurasiya) [inline src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: Can we not include the cache files?

**Review by Sachin-chaurasiya — CHANGES_REQUESTED** [review]
Some initial comments

File: src/plays/web3-staking-dapp/Web3StakingDapp.js (hunk 5 above)
Comment (nirban256) [reply src/plays/web3-staking-dapp/Web3StakingDapp.js:16]: So I have to create a file called constant,js and then I have to keep the contract address there?
This is what you are saying

Comment (Sachin-chaurasiya) [reply src/plays/web3-staking-dapp/Web3StakingDapp.js:16]: Yes, Exactly.

Comment (nirban256) [reply src/plays/web3-staking-dapp/Web3StakingDapp.js:16]: ok

File: src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json (hunk 9 above)
Comment (nirban256) [reply src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: can you tell me the what to write in the gitignore in order to remove the cache file?
I have tried with ./smart-contract/cache/solidity-files-cache.json

File: src/plays/web3-staking-dapp/Web3StakingDapp.js (hunk 8 above)
Comment (nirban256) [reply src/plays/web3-staking-dapp/Web3StakingDapp.js:47]: then it will show a warning i think

File: src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json (hunk 9 above)
Comment (Sachin-chaurasiya) [reply src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: Can you try `cache/*`?

File: src/plays/web3-staking-dapp/Web3StakingDapp.js (hunk 8 above)
Comment (Sachin-chaurasiya) [reply src/plays/web3-staking-dapp/Web3StakingDapp.js:47]: Which warning?

File: src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json (hunk 9 above)
Comment (nirban256) [reply src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: it didn't worked

Comment (Sachin-chaurasiya) [reply src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: @nirban256 , if you have already pushed the changes and they trying to ignore that files in gitignore then it won't work.

Comment (nirban256) [reply src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: @Sachin-chaurasiya I made another commit after editing the readme file and then it failed the test

Comment (nirban256) [reply src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: I don't know the exact reason for it

Comment (atapas) [reply src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: I've re-run the failed test. It should pass now. Can you please take care of the comments making sure they are done?

Comment (nirban256) [reply src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: @atapas I could not add the cache folder in the gitignore properly this is the only issue that is left. Can you tell me how to add the folder so that git does not track it?

**Comment by nirban256** [comment]
If all the conditions are met then please merge the pr

**Review by Sachin-chaurasiya — CHANGES_REQUESTED** [review]
@nirban256 Few more comments and remove the `src/plays/web3-staking-dapp/styles.css` as it's empty.

File: src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json (hunk 9 above)
Comment (nirban256) [reply src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:1]: ok

File: src/plays/web3-staking-dapp/smart-contract/scripts/deploy.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  block = await provider.getBlock(receipt.blockNumber);
+  newUnlockDate = block.timestamp - (60 * 60 * 24 * 1000);
+  await staking.connect(signer1).changeUnlockPeriod(4, newUnlockDate);
+
+}
+
+// We recommend this pattern to be able to use async/await everywhere
+// and properly handle errors.
+main().catch((error) => {
+  console.error(error);
```
[the remark this answers is not in the document]
Comment (nirban256) [reply src/plays/web3-staking-dapp/smart-contract/scripts/deploy.js:53]: having this statement will not create any problem because this file will not run when the play will run, it's work is done.

**Comment by Sachin-chaurasiya** [comment]
@nirban256 , Build is failing.

image

**Comment by nirban256** [comment]
> @nirban256 , Build is failing, As we have deleted the styles.css and we have to remove the reference of it as well.
> 
> image

removed the line @Sachin-chaurasiya

**Review by Sachin-chaurasiya — APPROVED** [review]
Great work @nirban256

**Comment by nirban256** [comment]
thanks @Sachin-chaurasiya

**Review by atapas — APPROVED** [review]
Approving
